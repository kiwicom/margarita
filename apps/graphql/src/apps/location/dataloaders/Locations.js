// @flow

import stringify from 'json-stable-stringify';
import qs from 'querystring';
import { OptimisticDataloader } from '@kiwicom/graphql-utils';

import fetch from '../../../services/fetch/tequilaFetch';
import { type Location, type ApiLocation } from '../Location';

export type Locations = $ReadOnlyArray<Location>;

export type LocationTypeInput =
  | 'STATION'
  | 'AIRPORT'
  | 'BUS_STATION'
  | 'CITY'
  | 'AUTONOMOUS_TERRITORY'
  | 'SUBDIVISION'
  | 'COUNTRY'
  | 'REGION'
  | 'CONTINENT'
  | 'SPECIAL';

export type LocationInput =
  | {|
      +term: string,
      +types?: LocationTypeInput[],
    |}
  | {|
      +code: string,
    |};

function sanitizeCity(location: ApiLocation) {
  if (location.city != null) {
    return {
      id: location.city.id,
      locationId: location.city.id,
      code: location.city.code,
      flagURL: null,
      slug: location.city.slug,
      name: location.city.name,
    };
  }
  return null;
}

function sanitizeLocations(locations: $ReadOnlyArray<ApiLocation>) {
  return locations.map(location => ({
    id: location.id,
    locationId: location.id,
    name: location.name,
    slug: location.slug,
    type: location.type,
    timezone: location.timezone,
    coordinates: {
      lat: location.location.lat,
      lng: location.location.lon,
    },
    country: {
      id: location.city?.country?.id ?? '',
      locationId: location.city?.country?.id,
      code: location.city?.country?.code,
      flagURL: null,
      slug: location.city?.country?.slug,
      name: location.city?.country?.name,
    },
    city: sanitizeCity(location),
  }));
}

const fetchLocations = async (params: $ReadOnlyArray<LocationInput>) => {
  const data = await Promise.all(
    params.map(param => {
      const locationsTypes = param.types ? param.types : [];

      // convert locations types into locations params
      const locationsParamTypes = locationsTypes.reduce((acc, type) => {
        return `${acc}&location_types=${type}`;
      }, '');

      if (param.term !== undefined) {
        return fetch(
          `/locations/query?${qs.stringify({
            term: param.term,
          })}${locationsParamTypes}`,
        );
      }

      return fetch(`/locations/id?id=${param.code}`);
    }),
  );
  return data.map(({ locations }) => sanitizeLocations(locations));
};

export default function locationsLoader() {
  return new OptimisticDataloader(
    (ids: $ReadOnlyArray<LocationInput>): Promise<Array<Location[] | Error>> =>
      fetchLocations(ids),
    {
      cacheKeyFn: stringify,
    },
  );
}
