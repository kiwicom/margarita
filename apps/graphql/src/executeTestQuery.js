// @flow

import { graphql as originalGraphQL } from 'graphql';

import schema from './Schema';
import createContext from './services/GraphQLContext';

/**
 * Executes GraphQL query using our schema. This function should
 * be used only during testing because it uses faked context.
 */
export default function executeTestQuery(
  query: string,
  variables: ?Object,
  acceptLanguage: ?string,
): Promise<Object> {
  return originalGraphQL(schema, query, null, createContext(), variables);
}
