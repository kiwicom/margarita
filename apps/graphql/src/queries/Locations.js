// @flow

import { GraphQLNonNull } from 'graphql';
import { type ConnectionArguments } from 'graphql-relay';
import {
  connectionFromArray,
  connectionArgs,
  connectionDefinitions,
} from '@kiwicom/graphql-utils';

import type { GraphqlContextType } from '../services/GraphQLContext';
import type { Location } from '../dataloaders/LocationsloaderTypes';
import GraphQLocation from '../types/output/Location';
import LocationsByTermInput from '../types/input/LocationsByTermInput';

const { connectionType: LocationsConnection } = connectionDefinitions({
  nodeType: GraphQLocation,
});

type Args = {|
  +input: { term: string },
  ...$Exact<ConnectionArguments>,
|};

const Locations = {
  name: 'Locations',
  type: LocationsConnection,
  description: 'Query for suggested locations based on incomplete names',
  args: {
    input: {
      type: GraphQLNonNull(LocationsByTermInput),
    },
    ...connectionArgs,
  },
  resolve: async (_: mixed, args: Args, { dataLoader }: GraphqlContextType) => {
    const { term } = args.input;
    const locations = await dataLoader.locations.load({
      term,
    });

    return connectionFromArray<Location>([...locations], args);
  },
};

export default Locations;
