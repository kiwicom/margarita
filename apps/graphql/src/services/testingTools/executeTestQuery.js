// @flow

import { graphql as originalGraphQL } from 'graphql';

import schema from '../../Schema';
import createContext from '../graphqlContext/GraphQLContext';

/**
 * Executes GraphQL query using our schema. This function should
 * be used only during testing because it uses faked context.
 */
export default function executeTestQuery(
  query: string,
  variables: ?Object,
): Promise<Object> {
  return originalGraphQL(schema, query, null, createContext(), variables);
}
