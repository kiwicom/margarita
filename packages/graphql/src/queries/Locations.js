// @flow

import { GraphQLString, GraphQLNonNull } from 'graphql';
import { type ConnectionArguments } from 'graphql-relay';
import {
  connectionFromArray,
  connectionArgs,
  connectionDefinitions,
} from '@kiwicom/graphql-utils';

import type { GraphqlContextType } from '../services/GraphQLContext';
import type { Location } from '../dataloaders/Locationsloader';
import GraphQLocation from '../types/output/Location';

const { connectionType: LocationsConnection } = connectionDefinitions({
  nodeType: GraphQLocation,
});

type Args = {|
  +term: string,
  ...$Exact<ConnectionArguments>,
|};

export default {
  name: 'Locations',
  type: LocationsConnection,
  description: 'Query for suggested locsation based on incomplete names',
  args: {
    term: {
      type: GraphQLNonNull(GraphQLString),
    },
    ...connectionArgs,
  },
  resolve: async (_: mixed, args: Args, { dataLoader }: GraphqlContextType) => {
    const { term } = args;
    const locations = await dataLoader.locations.load({
      term,
    });

    return connectionFromArray<Location>([...locations], args);
  },
};
