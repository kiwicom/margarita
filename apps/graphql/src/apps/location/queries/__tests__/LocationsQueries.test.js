// @flow

import { generateTestsFromFixtures } from '@kiwicom/test-utils';

import executeTestQuery from '../../../../services/testingTools/executeTestQuery';
import Locations from '../../__datasets__/locations';

fetch.mockResponses([JSON.stringify(Locations)]);

generateTestsFromFixtures(`${__dirname}/__fixtures__`, input =>
  executeTestQuery(input),
);
