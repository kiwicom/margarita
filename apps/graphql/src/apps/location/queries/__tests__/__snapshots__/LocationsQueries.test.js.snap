// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches expected output: LocationsByTerm.graphql 1`] = `
~~~~~~~~~~ INPUT ~~~~~~~~~~
{
  locationsByTerm(input: { term: "Prague" }) {
    edges {
      node {
        id
        slug
        name
        countryFlagURL
        city {
          id
          name
        }
        coordinates {
          lat
          lng
        }
      }
    }
  }
}

~~~~~~~~~~ OUTPUT ~~~~~~~~~~
{
  "data": {
    "locationsByTerm": {
      "edges": [
        {
          "node": {
            "id": "TG9jYXRpb246cHJhZ3VlX2N6",
            "slug": "prague-czechia",
            "name": "Prague",
            "countryFlagURL": "https://images.kiwi.com/flags/32x32/anywhere.png",
            "city": null,
            "coordinates": {
              "lat": 50.075538,
              "lng": 14.4378
            }
          }
        },
        {
          "node": {
            "id": "TG9jYXRpb246UFJH",
            "slug": "vaclav-havel-airport-prague-prague-czechia",
            "name": "Václav Havel Airport Prague",
            "countryFlagURL": "https://images.kiwi.com/flags/32x32/cz.png",
            "city": {
              "id": "TG9jYXRpb25BcmVhOnByYWd1ZV9jeg==",
              "name": "Prague"
            },
            "coordinates": {
              "lat": 50.100833,
              "lng": 14.26
            }
          }
        },
        {
          "node": {
            "id": "TG9jYXRpb246cHJhZ3VlLWNhc3RsZV9wb2k=",
            "slug": "prague-castle",
            "name": "Prague Castle",
            "countryFlagURL": "https://images.kiwi.com/flags/32x32/anywhere.png",
            "city": null,
            "coordinates": {
              "lat": 50.090811,
              "lng": 14.400523
            }
          }
        },
        {
          "node": {
            "id": "TG9jYXRpb246b2xkLXRvd24tc3F1YXJlLXByYWd1ZV9wb2k=",
            "slug": "old-town-square",
            "name": "Old Town Square, Prague",
            "countryFlagURL": "https://images.kiwi.com/flags/32x32/anywhere.png",
            "city": null,
            "coordinates": {
              "lat": 50.087462,
              "lng": 14.421211
            }
          }
        },
        {
          "node": {
            "id": "TG9jYXRpb246Q1otUFJBRy1QUkE2",
            "slug": "prague-prague-czechia",
            "name": "Prague ,Hradčanská",
            "countryFlagURL": "https://images.kiwi.com/flags/32x32/cz.png",
            "city": {
              "id": "TG9jYXRpb25BcmVhOnByYWd1ZV9jeg==",
              "name": "Prague"
            },
            "coordinates": {
              "lat": 50.096993,
              "lng": 14.402517
            }
          }
        },
        {
          "node": {
            "id": "TG9jYXRpb246UEZM",
            "slug": "florenc-bus-station-prague-czechia",
            "name": "Prague - Florenc Bus Station",
            "countryFlagURL": "https://images.kiwi.com/flags/32x32/cz.png",
            "city": {
              "id": "TG9jYXRpb25BcmVhOnByYWd1ZV9jeg==",
              "name": "Prague"
            },
            "coordinates": {
              "lat": 50.089525,
              "lng": 14.440954
            }
          }
        },
        {
          "node": {
            "id": "TG9jYXRpb246Q1otUFJBRy1QUkEy",
            "slug": "prague-liben-prague-czechia",
            "name": "Prague - Libeň",
            "countryFlagURL": "https://images.kiwi.com/flags/32x32/cz.png",
            "city": {
              "id": "TG9jYXRpb25BcmVhOnByYWd1ZV9jeg==",
              "name": "Prague"
            },
            "coordinates": {
              "lat": 50.100761,
              "lng": 14.502088
            }
          }
        },
        {
          "node": {
            "id": "TG9jYXRpb246Q1otUFJBRy1QUkFH",
            "slug": "prague-andel-prague-czechia",
            "name": "Prague Anděl",
            "countryFlagURL": "https://images.kiwi.com/flags/32x32/cz.png",
            "city": {
              "id": "TG9jYXRpb25BcmVhOnByYWd1ZV9jeg==",
              "name": "Prague"
            },
            "coordinates": {
              "lat": 50.068608,
              "lng": 14.404425
            }
          }
        },
        {
          "node": {
            "id": "TG9jYXRpb246Q1otUFJBRy1QUkEw",
            "slug": "prague-east-cerny-most-prague-czechia",
            "name": "Prague East (Černý Most)",
            "countryFlagURL": "https://images.kiwi.com/flags/32x32/cz.png",
            "city": {
              "id": "TG9jYXRpb25BcmVhOnByYWd1ZV9jeg==",
              "name": "Prague"
            },
            "coordinates": {
              "lat": 50.108898,
              "lng": 14.577435
            }
          }
        },
        {
          "node": {
            "id": "TG9jYXRpb246Q1otUFJBRy1QUkE0",
            "slug": "prague-roztyly-prague-czechia",
            "name": "Prague Roztyly",
            "countryFlagURL": "https://images.kiwi.com/flags/32x32/cz.png",
            "city": {
              "id": "TG9jYXRpb25BcmVhOnByYWd1ZV9jeg==",
              "name": "Prague"
            },
            "coordinates": {
              "lat": 50.037867,
              "lng": 14.476984
            }
          }
        }
      ]
    }
  }
}
`;
